### PR TITLE
Add Python Reflection Client

### PR DESCRIPTION
**Pull request by tomerv**
Implement ProtoReflectionDescriptorDatabase in Python to support
client-side reflection sevices.

See gRFC L95.

See also #29023.

@lidizheng

### CODE REVIEW

File: src/python/grpcio_tests/tests/reflection/_reflection_client_test.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    def testFindMethodError(self):
+        service_name = self._SERVICE_NAMES[0]
+        service_desc = self.desc_pool.FindServiceByName(service_name)
+
+        # FindMethodByName sometimes raises a KeyError, and sometimes returns None.
+        # See https://github.com/protocolbuffers/protobuf/issues/9592
+        with self.assertRaises(KeyError):
+            res = service_desc.FindMethodByName(_INVALID_SYMBOL_NAME)
+            if res is None:
+                raise KeyError()
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/reflection/_reflection_client_test.py:132]: Can you share more detail here? Are we going to remove this test, or what improvement did we add to make it compatible?

https://github.com/grpc/grpc/pull/29023#issuecomment-1060399713

Comment (tomerv) [reply src/python/grpcio_tests/tests/reflection/_reflection_client_test.py:132]: The original test was
```
        with self.assertRaises(KeyError):
            service_desc.FindMethodByName(_INVALID_SYMBOL_NAME)
```
I changed it to translate a `None` return value to a `raise KeyError`.

Comment (lidizheng) [reply src/python/grpcio_tests/tests/reflection/_reflection_client_test.py:132]: Thanks for the clear write-up in https://github.com/protocolbuffers/protobuf/issues/9592.

Checking other find* methods ([Python code](https://github.com/protocolbuffers/protobuf/blob/349d74d92e8268c41fbb726b0a8292d588c63e20/python/google/protobuf/descriptor_database.py)), they favor the raise behavior. Do you know how the C-Extension implementation will behave? Based on [code](https://github.com/protocolbuffers/protobuf/blob/2d69d44cc18beaa0a3a62d353f80aead669abb5c/python/google/protobuf/pyext/descriptor_database.cc#L113), their return type is `bool`.

Comment (tomerv) [reply src/python/grpcio_tests/tests/reflection/_reflection_client_test.py:132]: In the code you linked, the return type is `bool` but internally it calls a Python function (`PyObject_CallMethod(...)`). In this context we have a call chain that look like `Python -> C++ -> Python`, where the outer Python code is the one doing the call `desc_database.FindFileByName(...)` and the inner is the `FindFileByName` implementation. I'm not an expert on this, but I believe that if the inner Python code raises an exception then it is immediately caught by the outer Python code - it doesn't even do exception folding of that C++ function, just sort of steals the context.

If the exception is not caught by the outer Python code, it looks like this:
```
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    desc_pool.FindFileByName("foo")
KeyError: "Couldn't find file foo"
```
Note how all the traceback of the inner `C++ -> Python` is missing as result of this unusual flow.

I hope this answers your question.

**Review by lidizheng — APPROVED** [review]
LGTM.

**Comment by tomerv** [comment]
> LGTM.

Thanks! When can this be merged?

**Comment by lidizheng** [comment]
@gnossen PTAL.

@tomerv The policy was each PR requires 2 maintainers review.

**Review by gnossen — APPROVED** [review]

**Pull request merged**